Eliminate use of cmp in ocamltest

dra27 opened the pull request:
This splits off the ocamltest changes from #1380.

Note that I also realised that the `\r` processing was previously being done for all files - it's now only done for text files.

dra27 committed:
Eliminate use of cmp in ocamltest

Previously, ocamltest attempted to normalise the output of commands on
Windows to be Unix files. This fails on a Windows git checkout with the
normal core.autocrlf set to true.

The implementation has been revised not to use cmp and to strip \r from
the end of text files if every line ends in one.

shindere commented:
Do we really want to eliminate cmp completely from the picture?

I thought we would start by using it and fall back to something else
only if it reports a difference?

shindere commented:
David Allsopp (2017/10/12 05:53 -0700):
> This splits off the ocamltest changes from #1380.
> 
> Note that I also realised that the `\r` processing was previously
> being done for all files - it's now only done for text files.

Did it really happen that this removal of '\r' was invoked on binary
files?

shindere commented:
Btw: how about renaming bytes_to_ignore to bytes_to_skip?

I realise it was me who chose the original name. I was probably inspired
by cmp's man page but perhaps "skip" is more accurate than "ignore"?

@dra27 just discussing the idea, the effective change can be done later
I think.

dra27 commented:
@shindere - I mean that the previous version of *my* patch from #1380 incorrectly stripped `\r` for binary files (in a very specific case - you'd need it to be that every `\n` in the binary file was preceded by `\r`), not that what you had written before did! I fixed that when I split the code off to this GPR.

Eliminating cmp is definitely beneficial for Windows - it was outside the remit of that original GPR, but it seemed to be what was wanted (from @xavierleroy and @alainfrisch's comments). I'm always in favour of eliminating any native Windows program calling out to the shell, because the utilities typically don't exist.

To me ignore is more natural than skip - skip describes the physical operation you are doing to the channel, but ignore is what you are doing semantically (cf. all the `--ignore*` options to `git diff` and friends). Arguably the variable name should possible mention that fact that it's ignoring the first n bytes, but I think it's pretty clear as you named it originally!

xavierleroy commented:
Yes, I support the approach of comparing files for equality using Caml code instead of `cmp`, for better control of end-of-line processing, and also for speed.

dra27 commented on ocamltest/filecompare.ml:
```diff
@@ -53,22 +56,35 @@ type files = {
+    else raise Exit
+  in
+  let rec loop acc =
+    match input_line ic with
```
@damiendoligez wrote:
> I'm slightly nervous at the presence of both `bytes_to_ignore` and `input_line`. `bytes_to_ignore` exists because we want to compare binary executables, which are not text.
>
> `input_line` might fail on a 32-bit machine because of the string length limitation.
>
> More important, the automatic detection of line endings will be thrown off by binary files if they happen to contain some `\r\n` sequences as well as some `\n` characters on their own.

I don't *think* that the `input_line` logic can get confused here, because the file is opened using `open_in_bin` (so `\r` would just be returned in the string on all platforms). Note that `drop_cr` is not called for binary files (see a few lines later).

The string limit on 32-bit is potentially a problem, I agree, because some of the binary files can end up being somewhat large.

dra27 committed:
Compare binary files block by block

dra27 commented:
Extra code for comparing binary files block-by-block instead

shindere commented:
David Allsopp (2017/10/12 14:01 +0000):
> @shindere - I mean that the previous version of *my* patch from #1380
> incorrectly stripped `\r` for binary files (in a very specific case -
> you'd need it to be that every `\n` in the binary file was preceded by
> `\r`), not that what you had written before did!

Okay great, thanks. I was afraid I did yet another thing a wrong way. :)

> I fixed that when I split the code off to this GPR.

Thank you!

> Eliminating cmp is definitely beneficial for Windows - it was outside
> the remit of that original GPR, but it seemed to be what was wanted
> (from @xavierleroy and @alainfrisch's comments). I'm always in favour
> of eliminating any native Windows program calling out to the shell,
> because the utilities typically don't exist.

Sorry to come back to this. How about making this behaviour either
configurable or OS-dependent?

Rationale: given that most of the time the compared file will be the
same and that cmp is present under Unix it would make sense to me to use
it when it's possible.

I know that e.g. Frama-C use this strategy of calling cmp first and then
only calling diff when files are different. Sure, they are not
supporting Windows which makes the whole problem simpler.

> To me ignore is more natural than skip - skip describes the physical
> operation you are doing to the channel, but ignore is what you are
> doing semantically (cf. all the `--ignore*` options to `git diff` and
> friends). Arguably the variable name should possible mention that fact
> that it's ignoring the first n bytes, but I think it's pretty clear as
> you named it originally!

Yeah you got my point. I tried to find a name that the ignored bytes
were consecutive and at the beginning of the input. But if you like the
original name we can continue that way. Thanks.

shindere commented:
Xavier Leroy (2017/10/12 14:25 +0000):
> Yes, I support the approach of comparing files for equality using Caml
> code instead of `cmp`, for better control of end-of-line processing,
> and also for speed.

Caml code? For speed? ;-)

More seriously, what can be faster than a pure binary comparison which
will be sufficient most of the time?

nojb commented:
FWIW, launching processes can be quite slow on Cygwin.

shindere commented:
David Allsopp (2017/10/12 07:35 -0700):
> dra27 commented on this pull request.
> 
> 
> 
> > @@ -53,22 +56,35 @@ type files = {
>    output_filename : string;
>  }
>  
> -let unixify_end_of_lines filename =
> -  let commandline = "sed -i -e s/\\r$// " ^ filename in
> -  let status = Run_command.run_commandline commandline in
> -  (status, commandline)
> +let read_file bytes_to_ignore filetype fn =
> +  let ic = open_in_bin fn in
> +  seek_in ic bytes_to_ignore;
> +  let drop_cr s =
> +    let l = String.length s in
> +    if l > 0 && s.[l - 1] = '\r' then String.sub s 0 (l - 1)
> +    else raise Exit
> +  in
> +  let rec loop acc =
> +    match input_line ic with
 
It looks a bit weird to me to use input_line on a file opened with
open_in_bin, no?

alainfrisch commented:
> More seriously, what can be faster than a pure binary comparison which will be sufficient most of the time?

Not spawning an external process.

On a related topic, I observed recently that reading characters one by one with input_char can be quite slow compared to reading by chunks.  I'm wondering whether implementing the buffering of Pervasives channels in OCaml rather than in C wouldn't be faster.

shindere commented:
David Allsopp (2017/10/12 15:21 +0000):
> Extra code for comparing binary files block-by-block instead

Yeah, this looks much better to me, thanks!

Now we could even implement the two stages comparison I mentionned, if
so desired.

xavierleroy commented:
> More seriously, what can be faster than a pure binary comparison which will be sufficient most of the time?

Not spawning a `cmp` process to do the job.  I agree the time spend actually reading and comparing the files is small in either case, especially since these files are usually small.  But fork-exec / spawn is not that fast, esp. in Win32, and even worse in Cygwin.

damiendoligez commented on ocamltest/filecompare.ml:
```diff
@@ -69,15 +68,56 @@ let read_file bytes_to_ignore filetype fn =
+    else
+      read (pos + bytes_read)
+  in
+  let bytes_read = block_size - read 0 in
```
I find it rather strange that `read` returns the number of bytes **not** read, and this double-subtraction business is a bit hard to read. Why not just return `pos + bytes_read` at line 90, and replace line 94 with `let bytes_read = read 0` ?

I think this would be even more natural:
```
  let rec read pos =
    let bytes_read = input ic buf pos (block_size - pos) in
    let newpos = pos + bytes_read in
    if bytes_read = 0 || newpos = blocksize then
      newpos
    else
      read newpos
  in
  let bytes_read = read 0 in
```

gasche commented:
Should we postpone this PR to 4.07? The other option is to have Damien's readability concern addressed and have him approve the PR -- and give his opinion on including this in 4.06.

dra27 commented:
Give me a few minutes and I'll push some code addressing @damiendoligez's comments

dra27 commented:
Rebased and adopted @damiendoligez's clearer code.

gasche commented:
@damiendoligez, should we merge this in 4.06?

damiendoligez commented:
@gasche, yes we should merge it into 4.06 because it fixes the handling of \r.

gasche commented:
Merged, cherry-picked in 4.06 ( 6df5a2028c1219383e4670ab0b8234f190ba9545 ).
